docs/testing_image.md

```diff
@@ -0,0 +1,254 @@
+* Parse yaml files to find arguments and environment variables to inject.
+* Guarantee completion within a specified duration.
+* Copy results to a known prefix.
+
+In order to address this we will:
+
+* Expect the image to adhere to a particular contract in terms of folder
+  structure:
+  - `/incoming/run/run-metadata.yaml  # job name, tarball location, author, commit`
+  - `/incoming/provider/provider-config.yaml  #  credentials, project name, etc`
+  - `/incoming/export/export-config.yaml  # credentials, export prefix for copying results`
+* Copy binaries into these images that find and respond to the state of these
+  folders.
+
+Note that `provider-config.yaml` may not be used for tests that do not need to
+deploy/access a remote cluster.
+
+## Logic common to each class
+
+Each class will define an entry point. Every job based on this class will invoke
+this entry point. The entry point will control things common between jobs.
+For example all unit test jobs run `hack/test-go.sh` which always runs `go test`
+and produces junit.xml results; and all e2e jobs call `hack/e2e.go` to deploy a
+cluster and call `ginkgo e2e.test`, etc.
+
+## Logic unique to each job
+
+Each job will:
+* use common logic configure additional arguments and variables
+  - `/incoming/job/job-config.yaml  # Common job configuration (arguments/variables)`
+  - All the incoming files common to all jobs.
+* call the entrypoint
+  - using common logic to timeout after a specified duration.
+* use common logic to upload results if requested
+* exit 0 when everything passes otherwise non-zero
+
+
+# Initial classes
+
+Initially we will define seven job classes (and an image for each):
+
+1. `build` - build kubernetes
+2. `verify` - run verification scripts
+3. `unit` - run unit tests
+4. `integration` - run integration tests
+5. `e2e` - run e2e tests
+6. `node-e2e` - run node e2e test
+7. `kubemark` - run kubemark tests
+
+
+## build image
+
+This image will build kubernetes on one or more platforms using a
+`hack/build` script. This is analogous to `kubernetes-build`.
+
+This script will replace myriad other pieces involved in building:
+* `hack/e2e-internal/build-release.sh`
+* `hack/build-go.sh`
+* `hack/build-cross.sh`
+* `hack/build-ui.sh`
+* `hack/jenkins/build.sh`
+* `go run hack/e2e.go --build`
+
+The functionality in these scripts will be migrated into `hack/build`.
+We will update the scripts to first become wrappers for `hack/build`.
+Then we will deprecate the scripts. Then we will delete them.
+
+
+### build job-config
+
+This will control:
+* the subset of components to build: ui, server, client, tests, for exampl
+* the platforms to build each component
+
+This will default to a fast build.
+
+## verify image
+
+This image will run one or more verify scripts using the `hack/verify`. This
+is analogous to `kubernetes-verify`.
+
+The myriad of verify-* scripts in `hack` will move to `hack/verify-scripts/`.
```

Comment:
In the test-infra repo we moved them into `verify`. Ideally there is nothing in `hack`. It may also be worth splitting them into their own repo if we start breaking up the kubernetes repo, so that we don't have to keep all versions up to date.